### ISSUE TITLE
[公众号]Python中怎么对序列数据进行one-hot（独热）编码

### ISSUE THREAD

**Issue by briantmali** [briantmali/MachineLearning#9]
Python中怎么对序列数据进行one-hot（独热）编码

机器学习算法不能直接应对类别数据。
必须类别数据转换成数字。
上面的表述也同样适用于处理序列数据分类类型问题，比如当你想用诸如长短期记忆循环神经网络这样的深度学习方法的时候。
![image](https://user-images.githubusercontent.com/22410568/28397965-d1c35ce6-6d3f-11e7-9e45-009899873f2f.png)
How to One Hot Encode Sequence Classification Data in Python
Photo by Elias Levy, some rights reserved.
本文我们想分享如何将输入序列数据或者输出序列数据转换成one-hot编码，然后将one-hot编码用于深度学习中的序列分类问题。本文中我们将使用python来进行讲解。

阅读完本文，希望你能了解如下内容：
	什么是整数编码，什么是one-hot编码，为什么机器学习里需要这些编码
	怎么裸用来python实现计算整数编码和one-hot编码
	怎么用scilit-learn和Keras库来自动对你的序列数据进行编码

概述

本文分成4部分：
1.	什么是one-hot编码
2.	手动实现one-hot编码
3.	Scikit-learn实现one-hot编码
4.	Keras实现one-hot编码

什么是one-hot编码
one-hot编码是类别变量的表现形式，具体编码使用二进制向量。
进行one-hot编码就是用二进制向量来表现类别变量。
首先需要将类别值跟整数对应起来，也就是将类别值映射成整数。
然后将每个整数值表示成二进制向量。这个二进制向量中数值的个数（向量维数）等于所有类别的个数。二进制向量中所有值为0， 除了对应类别整数的座次（index）那一位被设为1。（抱歉，找不到index这个词通俗易懂的说明，只好借用梁山好汉的座次来勉强应对。）

One-hot编码的实例
让我们举个具体的例子。
假设我们有个由‘红’标签和‘绿’标签混合构成的序列。

整数编码
我们可以设定：
	红——整数0
	率——整数1
只要我们保证上边的对应关系不变，那么这个红绿的编码就叫整数编码。
上述的对应关系不变，那么我们以后就能从数字找到对应的标签。我们做预测时都要进行相同的处理。

One-hot编码
有了上面的整数编码，我们就能够进一步用二进制向量来表示每个整数值。上面的例子中，我们有两个类别，那么我们的二进制向量就是长度为2， 对应两个可能的整数值。
红标签编码成整数0， 可以表示为二进制向量[1, 0]。
	红——[1, 0]
这里座次0被标记为1。下一个绿标签编码成整数1， 那么可以表示为二进制向量[0, 1]。
	绿——[0,1]
这里座次1被标记为1。

比如我们有如下序列

'红', '红', '绿'

使用整数编码是，表示如下

0,0,1

使用one-hot编码的话表示成
[1,0]
[1,0]
[0,1]

为什么要用one-hot编码
One-hot编码让类别数据的表示比原来更有表现力。
许多机器学习算法不能直接处理类别数据。类别必须转换成数字。类别的输入变量和输出变量都要求转换。

我们要直接使用整数编码的话，需要针对具体问题进行调整。这种方式对于原本类别直接就具有自然的顺序关系的一些问题适用，这时类别转换成的整数也有顺序关系。比如温度标签：冷，暖，热就是如此。

但是也有些问题，标签之间本身没有任何顺序关系，这样数字转换后人为增加的顺序关系可能会对机器学习方法解决问题产生损害。具体的例子比如标签：狗，猫

后者的情况下，我们希望给整个网络更强的表现力，来为每个可能的标签学习出一个类似概率的数字。这一点让针对问题建模的过程更容易。当one-hot编码被用于输出变量时，它可以提供一个比单纯一个标签更细致的预测结果。

手动实现one-hot编码
这个例子中，我们假设我们有一个全是字母的字符串示例，但是这个示例序列不能涵盖所有可能出现的情况。

我们使用下面的字符串作为输入序列：
Hello world
我们假设所有可能的输入的世界里，全部都是小写字母和空格，没有任何其他符号。我们用这样的设定来展示怎么样搞定我们自己的one-hot编码。

代码如下：
`from numpy import argmax
# define input string
data = 'hello world'
print(data)
# define universe of possible input values
alphabet = 'abcdefghijklmnopqrstuvwxyz '
# define a mapping of chars to integers
char_to_int = dict((c, i) for i, c in enumerate(alphabet))
int_to_char = dict((i, c) for i, c in enumerate(alphabet))
# integer encode input data
integer_encoded = [char_to_int[char] for char in data]
print(integer_encoded)
# one hot encode
onehot_encoded = list()
for value in integer_encoded:
	letter = [0 for _ in range(len(alphabet))]
	letter[value] = 1
	onehot_encoded.append(letter)
print(onehot_encoded)
# invert encoding
inverted = int_to_char[argmax(onehot_encoded[0])]
print(inverted)`

上面的代码运行时先打印输入字符串。

对于所有可能输入的字符，生成一个字符值到整数值的一一对应关系也就是映射。然后用这个映射来对输入字符串进行编码。可以看到输入的第一个字母h被编码成7，也就是说h在所有可能的输入值（本例中为字符表）的排位座次为7。

这之后，整数编码被转换成one-hot编码。一次完成一个编码转换。生成一个字符表长度的全0值list，这样就可以表示任何预想内的字符。

下一步，某个具体字符的座次被标记为1,。我们可以看到第一个字母h的整数编码为7，表示为一个长为27，第7位设为1的二进制向量。

最后，我们验证一下从编码到标签的转换，并打印结果。我们用Numpy里的argmax()函数来定位二进制向量里最大的值（也就是1），用这个座次排位的整数值反向从字母表查找对应字符。

注意：下面的输出为了可读性调整了格式
hello world
 
[7, 4, 11, 11, 14, 26, 22, 14, 17, 11, 3]
 
[[0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]
 
h

现在我们从0开始实现了一下one-hot编码。下面让我们用scikit-learn来实现自动映射。

One-hot编码的scikit-learn实现
在本例中，我们假设有下面要讲到的输出序列只包含3个标签：
"cold"
"warm"
"hot"
一个10步长的序列示例如下所示：
cold, cold, warm, cold, hot, hot, warm, cold, warm, hot
首先需要一个整数编码例如1,2,3，然后把整数编码转成one-hot编码，也就是具有三个值的二进制向量，例如[1,0,0]。

上面10步长的序列中三个标签都出现了，我们用自动方法定义标签到整数的映射，以及从整数到二进制向量的映射。

本例中，我们使用scikit-learn中的encoders。具体来说就是用LabelEncoder来生成标签的整数编码，用OneHotEncoder来生成整数的one-hot编码。

代码如下：
`
from numpy import array
from numpy import argmax
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
# define example
data = ['cold', 'cold', 'warm', 'cold', 'hot', 'hot', 'warm', 'cold', 'warm', 'hot']
values = array(data)
print(values)
# integer encode
label_encoder = LabelEncoder()
integer_encoded = label_encoder.fit_transform(values)
print(integer_encoded)
# binary encode
onehot_encoder = OneHotEncoder(sparse=False)
integer_encoded = integer_encoded.reshape(len(integer_encoded), 1)
onehot_encoded = onehot_encoder.fit_transform(integer_encoded)
print(onehot_encoded)
# invert first example
inverted = label_encoder.inverse_transform([argmax(onehot_encoded[0, :])])
print(inverted)
`

例子中首先输出标签序列。跟着是标签的整数编码，最后是one-hot编码。
训练数据中包含了所有可能的标签的集合（set），这样我们就能根据整数编码和one-hot编码的转换完成一个标签到编码的完整映射。
OneHotEncoder 会默认返回一个更有效的稀疏编码。这对某些应用可能不适用，比如Keras深度学习库。在本例中，我们参数中设定sparse=False，不返回稀疏返回值类型。

如果我们得到这个3值one-hot编码的预测值，我们可以很容易反过来得到原始标签。
我们先用numpy的argmax（）函数来定位向量中最大值的座次排位，用这个排位号作为LabelEncoder 的输入，计算一个反向转换，来得到它的文本标签。

例子的最后将第一个one-hot编码还原成标签‘cold’。

例子的输出如下（下面的输出为了可读性调整了格式）：
['cold' 'cold' 'warm' 'cold' 'hot' 'hot' 'warm' 'cold' 'warm' 'hot']

[0 0 2 0 1 1 2 0 2 1]

[[ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 0.  0.  1.]
 [ 1.  0.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  0.  1.]
 [ 1.  0.  0.]
 [ 0.  0.  1.]
 [ 0.  1.  0.]]

['cold']

Keras的one-hot实现
有的问题中可能已经有了整数编码。
这时可以先进行编码调整，然后直接用调整过的整数。相反，也可以直接用整数值的one-hot编码。这里如果整数值直接没有真正的顺序关系，只不过是标签必须有个数值代表的话，就需要仔细斟酌一下上边的两种做法。

Keras提供一个函数叫to_categorical()，可以用来做整数的one-hot编码。
本例中，我们有4个整数值[0,1,2,3]，我们有下面由这4个整数构成的，长度为10的输入序列。
data = [1, 3, 2, 0, 3, 2, 2, 1, 0, 1]
上面的序列用到了4个整数中所有的数值，所以我们可以直接使用categorical()函数。相反，如果序列中起始值有0，但是没有出现所有的可能值，我们需要预先在函数中指定参数类别数目，比如to_categorical(num_classes=4)

下面给出完整的示例代码。
`
from numpy import array
from numpy import argmax
from keras.utils import to_categorical
# define example
data = [1, 3, 2, 0, 3, 2, 2, 1, 0, 1]
data = array(data)
print(data)
# one hot encode
encoded = to_categorical(data)
print(encoded)
# invert encoding
inverted = argmax(encoded[0])
print(inverted)
`

示例中先定义输入序列，并打印输出。

然后将整数编码为二进制向量并打印输出。可以看到，第一个整数1的编码为[0,1,0,0]，符合预期。

再后来，我们用Numpy里的argmax()函数反转序列中第一个值的编码，可以看到第一个整数的结果是1。

[1 3 2 0 3 2 2 1 0 1]

[[ 0.  1.  0.  0.]
 [ 0.  0.  0.  1.]
 [ 0.  0.  1.  0.]
 [ 1.  0.  0.  0.]
 [ 0.  0.  0.  1.]
 [ 0.  0.  1.  0.]
 [ 0.  0.  1.  0.]
 [ 0.  1.  0.  0.]
 [ 1.  0.  0.  0.]
 [ 0.  1.  0.  0.]]

1

推荐阅读
这里列出一些链接，可以进一步了解one-hot编码相关内容。
•	What is one hot encoding and when is it used in data science? on Quora
•	OneHotEncoder scikit-learn API documentation
•	LabelEncoder scikit-learn API documentation
•	to_categorical Keras API documentation
•	Data Preparation for Gradient Boosting with XGBoost in Python
•	Multi-Class Classification Tutorial with the Keras Deep Learning Library
总结
这里我们介绍了python中如何将类别序列用one-hot编码表示，为进一步进行深度学习做准备。

具体而言，应该可以了解如下内容：
	什么是整数编码，什么是one-hot编码，为什么在机器学习中有这样的编码要求。
	怎么裸用python来实现整数编码和one-hot编码。
	如何使用scikit-learn和Keras库来自动对序列进行编码。

序列数据准备中如果有什么问题可以随时交流。

原文链接：
http://machinelearningmastery.com/how-to-one-hot-encode-sequence-data-in-python/
作者：
Jason Brownlee
